From: Simon McVittie <simon.mcvittie@collabora.co.uk>
Date: Tue, 11 Feb 2014 14:22:04 +0000
Subject: [PATCH] gdbus tests: wait up to 60s for gdbus-testserver to take
 its bus name

Previously, we waited up to 0.5s, but that can fail on slow architectures
like ARM; now we wait up to 60s in 0.1s increments.

Bug: https://bugzilla.gnome.org/show_bug.cgi?id=724113
Forwarded: yes
---
 gio/tests/gdbus-connection-loss.c | 28 +++++++++++++++++++++++++---
 gio/tests/gdbus-threading.c       | 27 ++++++++++++++++++++++-----
 2 files changed, 47 insertions(+), 8 deletions(-)

--- a/gio/tests/gdbus-connection-loss.c
+++ b/gio/tests/gdbus-connection-loss.c
@@ -56,11 +56,20 @@
   return FALSE; /* remove source */
 }
 
+static gboolean
+give_up (gpointer data)
+{
+  g_error ("%s", (const gchar *) data);
+  g_return_val_if_reached (TRUE);
+}
+
 static void
 test_connection_loss (void)
 {
   GDBusProxy *proxy;
   GError *error;
+  gchar *name_owner;
+  guint id;
 
   error = NULL;
   proxy = g_dbus_proxy_new_sync (c,
@@ -73,6 +82,22 @@
                                  &error);
   g_assert_no_error (error);
 
+  id = g_timeout_add_seconds (60, give_up,
+      "waited more than ~ 60s for gdbus-testserver to take its bus name");
+
+  while (TRUE)
+    {
+      name_owner = g_dbus_proxy_get_name_owner (proxy);
+
+      if (name_owner != NULL)
+        break;
+
+      g_main_context_iteration (NULL, TRUE);
+    }
+
+  g_source_remove (id);
+  g_free (name_owner);
+
   error = NULL;
   g_dbus_proxy_call (proxy,
                      "Sleep",
@@ -124,9 +149,6 @@
   g_assert (g_spawn_command_line_async (path, NULL));
   g_free (path);
 
-  /* wait for the service to come up */
-  usleep (500 * 1000);
-
   /* Create the connection in the main thread */
   error = NULL;
   c = g_bus_get_sync (G_BUS_TYPE_SESSION, NULL, &error);
--- a/gio/tests/gdbus-threading.c
+++ b/gio/tests/gdbus-threading.c
@@ -415,6 +415,13 @@
     }
 }
 
+static gboolean
+give_up (gpointer data)
+{
+  g_error ("%s", (const gchar *) data);
+  g_return_val_if_reached (TRUE);
+}
+
 static void
 test_method_calls_in_thread (void)
 {
@@ -422,6 +429,7 @@
   GDBusConnection *connection;
   GError *error;
   gchar *name_owner;
+  guint id;
 
   error = NULL;
   connection = g_bus_get_sync (G_BUS_TYPE_SESSION,
@@ -439,8 +447,20 @@
                                  &error);
   g_assert_no_error (error);
 
-  name_owner = g_dbus_proxy_get_name_owner (proxy);
-  g_assert_cmpstr (name_owner, !=, NULL);
+  id = g_timeout_add_seconds (60, give_up,
+      "waited more than ~ 60s for gdbus-testserver to take its bus name");
+
+  while (TRUE)
+    {
+      name_owner = g_dbus_proxy_get_name_owner (proxy);
+
+      if (name_owner != NULL)
+        break;
+
+      g_main_context_iteration (NULL, TRUE);
+    }
+
+  g_source_remove (id);
   g_free (name_owner);
 
   test_method_calls_on_proxy (proxy);
@@ -598,9 +618,6 @@
   g_assert (g_spawn_command_line_async (path, NULL));
   g_free (path);
 
-  /* wait for the service to come up */
-  usleep (500 * 1000);
-
   /* Create the connection in the main thread */
   error = NULL;
   c = g_bus_get_sync (G_BUS_TYPE_SESSION, NULL, &error);
